### PR TITLE
Add with-browser Docker image

### PR DESCRIPTION
**Pull request by The-9880**
Issue: https://github.com/grafana/synthetic-monitoring/issues/71
Adds a 3rd stage to the Dockerfile to build the sm-agent image from an Alpine base, installing `chromium-switshader`.

The Drone CI pipeline is also updated to publish new image versions (tagged `*-with-browser`) corresponding to the new build target:
* `latest-with-browser`
* `<semver-long>-with-browser`

We should probably prevent this from publishing until we have some logic in the agent that can leverage it.

### CODE REVIEW

File: scripts/configs/drone/main.jsonnet
Comment (roobre) [inline scripts/configs/drone/main.jsonnet:99]: What does this build arg do? I'm not finding it in the dockerfile 🤔

File: scripts/configs/drone/main.jsonnet
Comment (roobre) [inline scripts/configs/drone/main.jsonnet:108]: Not sure I get what's going on here. Why are we setting `tag` to a string with spaces and parentheses? If this isn't the image tag I'd take this chance to add a comment here.

File: scripts/configs/drone/main.jsonnet
Comment (roobre) [inline scripts/configs/drone/main.jsonnet:197]: I'd suggest to use a shorter `-browser` suffix instead of `-with-browser`, just for the sake of brevity. I think the short version does not leave much room for ambiguity. WDYT?

File: Dockerfile
Comment (roobre) [inline Dockerfile:23]: I think we're mixing two substantially different linux flavors here (alpine and debian) and I'm afraid of unforeseen consequences. Most prominently, the libc implementation is different, which may cause a binary to work in one but not the other. Since the binary is built outside of docker with many many layers of Makefiles, I don't know if it is being built with `CGO_ENABLED=0` (hope/think it is), otherwise it will break on alpine for sure. 

We're also copying the ca-certificates from debian into alpine, which should be okay... but looks strange.

If you have manually tested this works, I wouldn't block the PR for this, but I'd strongly suggest to make a follow-up PR migrating the base image to alpine. We've been looking for an excuse to do this to lower our number of CVEs, so this is an excuse as good as any other.

File: Dockerfile
Comment (roobre) [inline Dockerfile:10]: This changes what is the "default" result of running `docker build` on this repo. I think it might be good to document these two targets and its differences in the `README`.

File: Dockerfile
Comment (roobre) [inline Dockerfile:28]: What are these two for?

File: Dockerfile
Comment (roobre) [inline Dockerfile:35]: With browser added in, we'll have the agent start the k6 process, which will itself start a chromium process, which will spawn a gazillion processes itself. I think this calls for adding `tini` to make sure all those children are properly terminated.

File: Dockerfile
Comment (roobre) [inline Dockerfile:25]: At this point we will be running everything as root. I think this can be left for a follow-up as well, but running a browser as root makes me a bit scared so we might want to avoid that.

**Review by roobre — COMMENTED** [review]
Left a few small comments, and a few bigger ones that we can leave for follow-up PRs if this one has been tested working.

File: scripts/configs/drone/main.jsonnet
Comment (The-9880) [reply scripts/configs/drone/main.jsonnet:99]: Thank you for catching this -- I started with an ARG-based solution first and then migrated to a built target one, so this needs to be removed

File: scripts/configs/drone/main.jsonnet
Comment (The-9880) [reply scripts/configs/drone/main.jsonnet:108]: Yes this isn't the image tag, I was following the existing use of the `tag` param in this jsonnet file. I will rename this in the new usages to avoid confusion

File: scripts/configs/drone/main.jsonnet
Comment (The-9880) [reply scripts/configs/drone/main.jsonnet:197]: I agree! I noticed other Docker images like that for `k6` were using `-with-browser`, but I don't mind shortening it for us

File: Dockerfile
Comment (The-9880) [reply Dockerfile:10]: Right. We have some `helper` make scripts on `docker build ...` which aren't using in any automations AFAICT. The result of those would build the new final stage instead of the old one so that is worth documenting as well but shouldn't impact any of our current pipelines I think.

Will update `README`!

File: Dockerfile
Comment (The-9880) [reply Dockerfile:28]: I wondered if providing an env var might be useful instead of relying on PATH to have `chromium`/assuming the binary name, for when sm-agent leverages it.

File: Dockerfile
Comment (The-9880) [reply Dockerfile:25]: 👍 I can add a user and operate the remainder of the target from there

File: Dockerfile
Comment (The-9880) [reply Dockerfile:23]: I believe it is being built with `CGO_ENABLED=0` (@mem to confirm if you have a second).

I built and ran this Dockerfile locally and it seems to run the agent binary, though I didn't actually allocate a workload to it so maybe I'll do that and report back.

In either case, using Debian or Alpine, I've noticed that attempting to run headless chromium would result in an error I'm still trying to debug: `Failed to connect to the bus: Failed to connect to socket /var/run/dbus/system_bus_socket: No such file or directory`. I used `chromiuim`/`chromium-shell` on Debian in that case and got the same issue as `chromium-swiftshader` on Alpine.

I'm still working on understanding that, wondering if you have any insight as you got this running for crocochrome :eyes:

File: scripts/configs/drone/main.jsonnet
Comment (The-9880) [reply scripts/configs/drone/main.jsonnet:197]: Well, normally I would want it to be consistent with other releases but because we build a longer tag name for our images, including the commit hash, I think I like keeping it shorter for this case

File: Dockerfile
Comment (The-9880) [reply Dockerfile:23]: If we land on agreeing that this is fine, I would for sure migrate the base to alpine as well.

File: Dockerfile
Comment (The-9880) [reply Dockerfile:28]: Will we add `xk6-browser` to the k6 setup for this? We may want to add `K6_BROWSER_HEADLESS` and `K6_BROWSER_ARGS` for that

File: Dockerfile
Comment (roobre) [reply Dockerfile:23]: I think that error message is benign, I also get that on crocochrome is my memory serves. Probably just one of the million things chrome is trying to run at startup.

File: Dockerfile
Comment (roobre) [reply Dockerfile]: `xk6-browser` is now part of `k6`, so it's already there :D

I believe it should already be looking for the correct path, but we definitely want to tweak `K6_BROWSER_ARGS`. For instance, `--disable-dev-shm-usage` is required for some webpages to not crash on a container.

File: Dockerfile
Comment (roobre) [reply Dockerfile:10]: README changes look good, thanks!

File: Dockerfile
Comment (roobre) [reply Dockerfile:26]: This might fail in some corner case configurations (e.g. if both user namespaces and setuid bits are disabled) but I think it should work good enough for now!

Comment (roobre) [reply Dockerfile:26]: Re-thinking this (sorry)... This might get us into problems where running the agent as non-root might cause some check types to fail. e.g. traceroute/ping checks require either elevated privileges or net_raw capabilities to be added to the binary.

It might be better for now to run everything as root, and open an issue to narrow-down the privileges without breaking agent checks.

**Review by roobre — APPROVED** [review]
If this passes some happy path tests, it looks good to me as a first iteration to unblock early adopters and internal tests.

I'd suggest to open a few follow-up issues to act on before we make this more widely available:
- Migrate base image to alpine to consolidate in one distro
- Pin and auto update chromium versions (Copy from https://github.com/grafana/crocochrome/pull/5 when it is merged)
- Run both agent and chromium as non-root, granting required capabilities (i.e. `net_raw`) to the agent binary.

File: scripts/configs/drone/main.jsonnet
Comment (roobre) [reply scripts/configs/drone/main.jsonnet]: Hmm... that's a good point, consistency is good... I don't see a clear winner but I'm still leaning towards `-browser` for short. Maybe we can ask the rest of the team to see if anyone has a strong opinion?

**Comment by roobre** [comment]
Forgot to mention, making sure that `--disable-dev-shm-usage` is passed to chromium is probably something we should figure out in this PR.

File: Dockerfile
Comment (The-9880) [reply Dockerfile:23]: I assigned some workloads to this and it seems to run fine on my linux VM -- I will resolve this thread, with the takeaway that we want to migrate the original base to alpine as well!

File: Dockerfile
Comment (The-9880) [reply Dockerfile]: Done!

File: Dockerfile
Comment (The-9880) [reply Dockerfile]: The K6 Dockerfile seems to do the following:
```
# no-sandbox chrome arg is required to run chrome browser in
# alpine and avoids the usage of SYS_ADMIN Docker capability
ENV K6_BROWSER_ARGS=no-sandbox
```

Should we be using no-sandbox as well? I'll add it tentatively to this PR, will re-request your review again before I merge anything.

File: scripts/configs/drone/main.jsonnet
Comment (roobre) [inline scripts/configs/drone/main.jsonnet:194]: As a small nit, is there a reason for this command using eval and curly braces?
On my PC, this seems to work:
```console
echo "latest-browser,$(./scripts/version)-browser"
```

**Review by roobre — APPROVED** [review]
Looking good!

File: scripts/configs/drone/main.jsonnet
Comment (The-9880) [reply scripts/configs/drone/main.jsonnet:194]: That should work - updating!

**Comment by The-9880** [comment]
Merging without re-approval as the only change was to sort out the `.drone.yml` conflict with main.

**Pull request merged**
